data/transactions/logic/eval.go

```diff
@@ -4764,56 +4766,45 @@ func opBase64Decode(cx *EvalContext) error {
 	cx.stack[last].Bytes = bytes
 	return nil
 }
-func hasDuplicateKeys(jsonText []byte) (bool, map[string]json.RawMessage, error) {
+
+func isPrimitiveJSON(jsonText []byte) (bool, error) {
 	dec := json.NewDecoder(bytes.NewReader(jsonText))
-	parsed := make(map[string]json.RawMessage)
 	t, err := dec.Token()
 	if err != nil {
-		return false, nil, err
+		return false, err
 	}
 	t, ok := t.(json.Delim)
 	if !ok || t.(json.Delim).String() != "{" {
-		return false, nil, fmt.Errorf("only json object is allowed")
-	}
-	for dec.More() {
-		var value json.RawMessage
-		// get JSON key
-		key, err := dec.Token()
-		if err != nil {
-			return false, nil, err
-		}
-		// end of json
-		if key == '}' {
-			break
-		}
-		// decode value
-		err = dec.Decode(&value)
-		if err != nil {
-			return false, nil, err
-		}
-		// check for duplicates
-		if _, ok := parsed[key.(string)]; ok {
-			return true, nil, nil
-		}
-		parsed[key.(string)] = value
+		return true, nil
 	}
-	return false, parsed, nil
+	return false, nil
 }
 
 func parseJSON(jsonText []byte) (map[string]json.RawMessage, error) {
-	if !json.Valid(jsonText) {
-		return nil, fmt.Errorf("invalid json text")
-	}
-	// parse json text and check for duplicate keys
-	hasDuplicates, parsed, err := hasDuplicateKeys(jsonText)
-	if hasDuplicates {
-		return nil, fmt.Errorf("invalid json text, duplicate keys not allowed")
+	// guard against primitives
+	isPrimitive, err := isPrimitiveJSON(jsonText)
+	if err != nil {
+		return nil, err
+	} else if isPrimitive {
+		return nil, fmt.Errorf("invalid json text, only json object is allowed")
 	}
+
+	// parse JSON with Algorand's standard JSON library
+	var parsed map[string]json.RawMessage
+	err = protocol.DecodeJSON(jsonText, &parsed)
+
 	if err != nil {
-		return nil, fmt.Errorf("invalid json text, %v", err)
+		// if the error was caused by duplicate keys
+		if strings.Contains(err.Error(), "cannot decode into a non-pointer value") {
```

Comment:
And worse, it's an awfully weird error for duplicate keys.

Reply:
It is related to the way it assigns into the interface. It tries to get an address in the interface to assign the value to and since the value is already assigned (and not a pointer I'd assume 🤷‍♂️ ) it throws an error. It's basically complaining about a collision.

Reply:
Make sure you have a unit test that confirms that way of catching works - wouldn't want to take an update of the `codec` package and have them change the string.

Reply:
Added a test to catch this.
https://github.com/algorand/go-algorand/pull/4062/files#diff-d6e7252ce42ce19fa92f80dd475dff3e611cb2f5c2876395cc851245cd018e6bR4920-R4928